lib/Conversion/TorchToTcp/Misc.cpp

```diff
@@ -81,10 +83,28 @@ class ConvertAtenBroadcastToOp : public OpConversionPattern<AtenBroadcastToOp> {
     SmallVector<Value> resultShape;
     for (int64_t i = 0; i < static_cast<int64_t>(newDimSizes.size()); ++i) {
       Value newDimSize = newDimSizes[i];
+
+      bool isNewDim = i < newLeadingDims;
+      // At the pytorch level it's possible to broadcast to a dynamic dimension;
+      // for example: `torch.broadcast_to(x, y.size())` where `y.size()` has
+      // dynamic shapes. When broadcasting to dynamic size, matchPattern fails
+      // to read an int out of the MLIR Value. In the "usual" case
+      // (`torch.broadcast_to(x, (3, 3))`) matchPattern will succeed.
       int64_t staticDimSize;
-      if (i < newLeadingDims ||
-          !matchPattern(newDimSize, m_TorchConstantInt(&staticDimSize)) ||
-          staticDimSize != -1) {
+      bool isDimDynamic =
+          !matchPattern(newDimSize, m_TorchConstantInt(&staticDimSize));
+      // lambda because staticDimSize will contain garbage in the case of a
+      // dynamic dim.
+      auto isDimBroadcasted = [i, &inputShape,
+                               newLeadingDims](int64_t dimSize) -> bool {
+        // pytorch defines "passing -1 as the size for a dimension means not
+        // changing the size of that dimension."
+        bool isDimSizePreserved = dimSize == -1;
+        bool doesDimChangeShape = dimSize != inputShape[i - newLeadingDims];
```

Comment:
This will result in accessing with negative indices when `i < newLeadingDim`, if this lambda is used separately. It is probably not happening now because this lambda is preceded by `isNewDim` condition in your check below.

Since, this is now separated out, I suggest adding the `i >= newLeadingDims` check here as well. Wdyt?

Reply:
How about we drop the lambda entirely, and do something like this instead, to handle both the negative indices issue and the garbage staticDimSize issue?
```cpp
bool isNewDim = i < newLeadingDims;
int64_t staticDimSize;
bool isDimDynamic = !matchPattern(newDimSize, m_TorchConstantInt(&staticDimSize));
bool isDimSizePreserved = isDimDynamic ? false : staticDimSize == -1;
bool doesDimSizeChange = isNewDim ? false : staticDimSize != inputShape[i - newLeadingDims];
if (isNewDim || isDimDynamic || (!isDimSizePreserved && doesDimSizeChange)) {
...
```

Reply:
`doesDimSizeChange` also requires a non dynamic dim. How about just a comment that says the order in which the booleans are checked matters here?

Reply:
huh I realized that and edited it, but I think GH doesn't send edit emails 😅 
```cpp
bool doesDimSizeChange = (isNewDim || isDimDynamic) ? false : staticDimSize != inputShape[i - newLeadingDims];
```